File: handloom-admin-frontend/src/features/orders/lib/refundAmount.ts
/**
 * Mirrors deriveRefundAmount in internal/service/refund_amount.go.
 *
 * The server derives the refund and ignores any amount a client sends, so this
 * is a preview and never an input. It exists because an admin should see what
 * leaves the account before authorising it, not after.
 *
 * Keep in sync with the Go function — it is the source of truth. The tests use
 * the same figures as TestDeriveRefundAmount for exactly that reason.
 */

export interface RefundableLine {
  id: string;
  unit_price: number;
  quantity: number;
  refunded_quantity: number;
}

export interface RefundableOrder {
  items: RefundableLine[];
  subtotal: number;
  discount_amount: number;
  tax_amount: number;
  total_amount: number;
}

/** Quantities to refund, keyed by order item id. */
export type RefundSelection = Record<string, number>;

export interface RefundPreview {
  total: number;

  /** The three terms behind the total, so the screen can show why it is that
   * number rather than only what it is. */
  breakdown: {
    lineValue: number;
    discount: number;
    tax: number;
    /** Returned only by the refund that clears the order, and zero until then. */
    shipping: number;
  };

  /** What was actually priced, after clamping. This is what must be submitted —
   * sending the raw input would post a quantity the preview never costed. */
  lines: { orderItemId: string; quantity: number; amount: number }[];

  /** True when this refund clears the last unrefunded unit — the one that earns
   * back the shipping and absorbs the rounding residual. */
  isFinal: boolean;
}

/** Units of each line already spoken for by refunds that have not failed. */
export type RefundClaims = Record<string, number>;

/** The subset of a refund this module needs to count what is already claimed. */
export interface ClaimingRefund {
  status: string;
  amount: number;
  items: { order_item_id: string; quantity: number }[];
}

/**
 * What an order's refunds already account for — settled or still in flight.
 * Mirrors claimedByLive in internal/service/refund_service.go, and the two must
 * agree: a failed refund returned nothing, so its units are free again.
 *
 * settledAmount is the payment's own figure, which wins when it is larger. A
 * settlement that half-completed leaves the rows ahead of the money or behind
 * it, and the refundable remainder has to be the pessimistic reading of both.
 */
export function claimedByLiveRefunds(
  refunds: ClaimingRefund[],
  settledAmount = 0
): { claims: RefundClaims; amount: number } {
  const claims: RefundClaims = {};
  let amount = 0;

  for (const refund of refunds) {
    if (refund.status === 'FAILED') continue;
    amount += refund.amount;
    for (const item of refund.items) {
      claims[item.order_item_id] = (claims[item.order_item_id] ?? 0) + item.quantity;
    }
  }

  return { claims, amount: Math.max(amount, settledAmount) };
}

/**
 * What a line has left to refund.
 *
 * refunded_quantity counts settled refunds only, so a refund still in flight is
 * invisible to it — exactly the window in which the same units could go back
 * twice. claims carries those units, and the larger of the two wins.
 */
export function unrefundedQuantity(line: RefundableLine, claims?: RefundClaims): number {
  const settled = line.refunded_quantity ?? 0;
  const claimed = claims?.[line.id] ?? 0;
  return Math.max(0, line.quantity - Math.max(settled, claimed));
}

// value's share of the whole for a line worth lineSubtotal, rounded half-up.
// Integer paise throughout, matching the Go arithmetic exactly.
function prorate(value: number, lineSubtotal: number, subtotal: number): number {
  if (value === 0 || subtotal === 0) return 0;
  // BigInt because the product leaves a double's exact-integer range before it
  // leaves Go's int64: the server would then round differently, and the admin
  // authorises whatever this returns. The result is a paise figure well inside
  // Number.MAX_SAFE_INTEGER, so converting back is safe.
  const scaled = BigInt(value) * BigInt(lineSubtotal) + BigInt(subtotal) / 2n;
  return Number(scaled / BigInt(subtotal));
}

export function previewRefund(
  order: RefundableOrder,
  selection: RefundSelection,
  priorRefunded: number,
  claims?: RefundClaims
): RefundPreview {
  const lines: { orderItemId: string; quantity: number; amount: number }[] = [];
  const remainingAfter = new Map<string, number>();
  let runningTotal = 0;
  let lineValue = 0;
  let discount = 0;
  let tax = 0;

  for (const item of order.items) {
    const left = unrefundedQuantity(item, claims);

    // Clamped rather than rejected: the stepper already caps the input, and a
    // preview that throws would leave the admin staring at a blank panel.
    const quantity = Math.min(Math.max(selection[item.id] ?? 0, 0), left);
    if (quantity === 0) continue;

    remainingAfter.set(item.id, left - quantity);

    const lineSubtotal = item.unit_price * quantity;
    const lineDiscount = prorate(order.discount_amount, lineSubtotal, order.subtotal);
    const lineTax = prorate(order.tax_amount, lineSubtotal, order.subtotal);
    const amount = lineSubtotal - lineDiscount + lineTax;

    lineValue += lineSubtotal;
    discount += lineDiscount;
    tax += lineTax;
    runningTotal += amount;
    lines.push({ orderItemId: item.id, quantity, amount });
  }

  const isFinal = lines.length > 0 && clearsOrder(order, remainingAfter, claims);
  let total = runningTotal;

  if (isFinal) {
    // Whatever is left of the order is what is left to refund — shipping and
    // rounding residual included. Deriving it this way is what makes the
    // refunds sum to the order exactly.
    total = order.total_amount - priorRefunded;
    const adjust = total - runningTotal;
    if (adjust !== 0) {
      lines[lines.length - 1].amount += adjust;
    }
  }

  // Whatever the clearing refund absorbs beyond the per-line terms is the
  // shipping plus the rounding residual — shown as shipping, which is what it
  // overwhelmingly is.
  const shipping = isFinal ? total - (lineValue - discount + tax) : 0;

  return { total, lines, isFinal, breakdown: { lineValue, discount, tax, shipping } };
}

// Whether nothing would be left unrefunded once the selected lines go back.
function clearsOrder(
  order: RefundableOrder,
  remainingAfter: Map<string, number>,
  claims?: RefundClaims
): boolean {
  return order.items.every(
    (item) => (remainingAfter.get(item.id) ?? unrefundedQuantity(item, claims)) === 0
  );
}

File: handloom-admin/internal/service/refund_amount.go
package service

import (
	"fmt"

	"github.com/handloom/admin/internal/domain"
	"github.com/handloom/admin/pkg/errors"
)

// refundBreakdown is what a requested set of lines is worth. The client sends
// lines and quantities; this is the only thing that decides money.
type refundBreakdown struct {
	Total int64
	Items []domain.RefundItem

	// IsFinal means this refund clears the order's last unrefunded unit: it earns
	// back the shipping and absorbs the rounding residual.
	IsFinal bool

	// The terms behind Total, so a screen can show why it is that number. Shipping
	// carries the rounding residual too, which is what it overwhelmingly is.
	LineValue int64
	Discount  int64
	Tax       int64
	Shipping  int64
}

// prorate returns value's share of the whole for a line worth lineSubtotal,
// rounded half-up. Integer paise throughout — no float ever touches money.
func prorate(value, lineSubtotal, subtotal int64) int64 {
	if value == 0 || subtotal == 0 {
		return 0
	}
	return (value*lineSubtotal + subtotal/2) / subtotal
}

// deriveRefundAmount computes the refund: each line less its share of the discount.
// Prices are tax-inclusive, so adding tax on top would refund GST twice. claimed
// counts what refunds already spoke for, settled or in flight.
func deriveRefundAmount(order *domain.Order, requested []domain.CreateRefundItemRequest, claimed map[string]int, priorRefunded int64) (*refundBreakdown, error) {
	if len(requested) == 0 {
		return nil, errors.BadRequest("A refund needs at least one line")
	}

	byID := make(map[string]*domain.OrderItem, len(order.Items))
	for i := range order.Items {
		byID[order.Items[i].ID] = &order.Items[i]
	}

	seen := make(map[string]bool, len(requested))
	remaining := make(map[string]int, len(requested))

	breakdown := &refundBreakdown{Items: make([]domain.RefundItem, 0, len(requested))}
	var runningTotal int64

	for _, req := range requested {
		item, ok := byID[req.OrderItemID]
		if !ok {
			return nil, errors.BadRequest(fmt.Sprintf("Order has no line %s", req.OrderItemID))
		}
		if seen[req.OrderItemID] {
			// Two entries for one line would each pass the remainder check on
			// their own and together exceed it.
			return nil, errors.BadRequest(fmt.Sprintf("Line %s appears twice", req.OrderItemID))
		}
		seen[req.OrderItemID] = true

		left := item.Quantity - unrefundedBaseline(item, claimed)
		if req.Quantity < 1 || req.Quantity > left {
			return nil, errors.BadRequest(fmt.Sprintf(
				"Line %s has %d unit(s) left to refund, asked for %d", req.OrderItemID, left, req.Quantity))
		}
		remaining[req.OrderItemID] = left - req.Quantity

		lineSubtotal := item.UnitPrice * int64(req.Quantity)
		// The refunded units' share of what this line was discounted, clamped because
		// this reads stored data rather than dividing one figure by one subtotal.
		lineDiscount := prorate(item.DiscountAmount, int64(req.Quantity), int64(item.Quantity))
		lineDiscount = min(max(lineDiscount, 0), lineSubtotal)
		// Tax-inclusive, so tax is read back out of the line's value rather than
		// added to it — extractTax is for the credit note to itemize.
		amount := lineSubtotal - lineDiscount
		lineTax := extractTax(amount)

		breakdown.LineValue += lineSubtotal
		breakdown.Discount += lineDiscount
		breakdown.Tax += lineTax
		runningTotal += amount
		breakdown.Items = append(breakdown.Items, domain.RefundItem{
			OrderItemID: item.ID,
			ProductID:   item.ProductID,
			ProductName: item.ProductName,
			Quantity:    req.Quantity,
			Amount:      amount,
			Restock:     req.Restock,
		})
	}

	breakdown.IsFinal = clearsOrder(order, remaining, claimed)
	breakdown.Total = runningTotal

	if breakdown.IsFinal {
		// What is left of the order is what is left to refund, shipping and residual
		// included. Deriving it this way is what makes the refunds sum to the order.
		breakdown.Total = order.TotalAmount - priorRefunded
		if adjust := breakdown.Total - runningTotal; adjust != 0 {
			last := &breakdown.Items[len(breakdown.Items)-1]
			last.Amount += adjust
		}
	}

	if breakdown.Total < 0 || priorRefunded+breakdown.Total > order.TotalAmount {
		return nil, errors.BadRequest("Refund would exceed the order total")
	}

	// Whatever the clearing refund absorbs beyond the per-line terms is the shipping
	// plus the residual. Zero until then, because the parcel still ships. Tax is not
	// subtracted here: it was never added to LineValue-Discount in the first place.
	if breakdown.IsFinal {
		breakdown.Shipping = breakdown.Total - (breakdown.LineValue - breakdown.Discount)
	}

	return breakdown, nil
}

// unrefundedBaseline is what a line is already spoken for before this request: what
// refunds claimed, or the settled counter when no claim is supplied.
func unrefundedBaseline(item *domain.OrderItem, claimed map[string]int) int {
	if claimed == nil {
		return item.RefundedQuantity
	}
	if taken := claimed[item.ID]; taken > item.RefundedQuantity {
		return taken
	}
	return item.RefundedQuantity
}

// clearsOrder reports whether nothing would be left unrefunded once the
// requested lines go back.
func clearsOrder(order *domain.Order, remainingAfter, claimed map[string]int) bool {
	for i := range order.Items {
		item := &order.Items[i]

		left, requested := remainingAfter[item.ID]
		if !requested {
			left = item.Quantity - unrefundedBaseline(item, claimed)
		}
		if left > 0 {
			return false
		}
	}
	return true
}
